docs/architecture/adr-001-node-key-refactoring.md

```diff
@@ -0,0 +1,140 @@
+
+type Node struct {
+	key           []byte
+	value         []byte
+	hash          []byte     // keep it in the storage instead of leftHash and rightHash
+	nodeKey       *NodeKey   // new field, the key in the storage
+	leftNodeKey   *NodeKey   // new field, need to store in the storage
+	rightNodeKey  *NodeKey   // new field, need to store in the storage
+	leftNode      *Node
+	rightNode     *Node
+	size          int64
+	leftNode      *Node
+	rightNode     *Node
+	subtreeHeight int8
+}
+```
+
+New tree structure
+
+```go
+type MutableTree struct {
+	*ImmutableTree                                     // The current, working tree.
+	lastSaved                *ImmutableTree            // The most recently saved tree.
+	unsavedFastNodeAdditions map[string]*fastnode.Node // FastNodes that have not yet been saved to disk
+	unsavedFastNodeRemovals  map[string]interface{}    // FastNodes that have not yet been removed from disk
+	ndb                      *nodeDB
+	skipFastStorageUpgrade   bool // If true, the tree will work like no fast storage and always not upgrade fast storage
+	
+	mtx sync.Mutex
+}
+```
+
+We will assign the `nodeKey` when saving the current version in `SaveVersion`. It will reduce unnecessary checks in CRUD operations of the tree and keep sorted the order of insertion in the LSM tree.
+
+### Migration
+
+We can migrate nodes through the following steps:
+
+- Export the snapshot of the tree from the original version.
+- Import the snapshot to the new version.
+	- Track the nonce for the same version using int32 array of the version length.
+	- Assign the `nodeKey` when saving the node.
+
+### Pruning
+
+We assume keeping only the range versions of `fromVersion` to `toVersion`. Refer to [this issue](https://github.com/cosmos/cosmos-sdk/issues/12989).
+
+Here we are introducing a new way how to get orphaned nodes which remove in the `n+1`th version updates.
+
+- Traverse the tree in-order way based on the root of `n+1`th version.
+- If we visit the lower version node, pick the node and don't visit further deeply. Pay attention to the order of these nodes.
+- Traverse the tree in-order way based on the root of `n`th version.
+- Iterate the tree until meet the first node among the above nodes(stack) and delete all visited nodes so far from `n`th tree.
+- Pop the first node from the stack and iterate again.
+
+If we assume `1 to (n-1)` versions already been removed, when we want to remove the `n`th version, we can just remove the above orphaned nodes.
+
+### Rollback
+
+When we want to rollback to the specific version `n`
+
+- Iterate the version from `n+1`.
+- Traverse key-value through `traversePrefix` with `prefix=bigendian(version)`.
+- Remove all iterated nodes.
+
+## Consequences
+
+### Positive
+
+Using the version and the path, we take advantage of data locality in the LSM tree. Since we commit the sorted data, it can reduce compactions and makes it easy to find the key. Also, it can reduce the key and node size in the storage.
```

Comment:
```suggestion
Using the version and a local nonce, we take advantage of data locality in the LSM tree. Since we commit the sorted data, it can reduce compactions and makes it easy to find the key. Also, it can reduce the key and node size in the storage.
```